### DISCUSSION TITLE
ConfirmListener#handleNack() can not be triggered

### DISCUSSION THREAD

**Discussion by leopardrota** [rabbitmq/rabbitmq-server#8395] [Other] [answered]
### Describe the bug

When a message is unable to reach any queue(exchange is right, routingKey is wrong), `ConfirmListener#handleNack()` can not be triggered and `ConfirmListener#handleAck()` is always be invoked.

### Reproduction steps

server: 3.8.3
com.rabbimq:amqp-client:5.16.0

1.channel.confirmSelect();
channel.addConfirmListener(new MyConfirmListener());

2.channel.basicPublish(exchange, routingKey, props, data); //exchange is right, routingKey is wrong

### Expected behavior

`ConfirmListener#handleNack()` can be triggered not `ConfirmListener#handleAck()` 

### Additional context

Currently, the message can not reach any queue, but `ConfirmListener#handleAck()` can be invoked. So where does the message go? Is it directly discarded?

**Comment by michaelklishin** [comment, marked as answer]
A couple of things.

1. Your expectation about when a [negative publisher acknowledgement (confirm)](https://rabbitmq.com/confirms.html#server-sent-nacks) is sent is off: it will not be sent when the message cannot be routed. For that, see [mandatory routing and returned messages](https://rabbitmq.com/publishers.html#unroutable). `basic.nack` will be sent when the queue process cannot accept an ingress message (due to a failure of some kind, a lack of quorum, things like that: nothing to do with routing, as this happens *after* the routing step)
2. RabbitMQ 3.8 has [reached end of life](https://rabbitmq.com/versions.html) in July 2022
